precovery/brute_force.py

```diff
@@ -0,0 +1,521 @@
+            fi.obscode, fi.mjd_utc, fi.healpix_id
+        ):
+            filtered_frames.append(f)
+
+    return filtered_frames
+
+
+def attribution_worker(
+    ephemeris,
+    observations,
+    tolerance=1 / 3600,
+    include_probabilistic=True,
+):
+    """
+    gather attributions for a df of ephemerides, observations
+
+    First filters ephemerides to match the chunked observations
+
+    """
+
+    attributions = []
+    # Create observer's dictionary from observations
+    observers = {}
+    for observatory_code in observations["observatory_code"].unique():
+        observers[observatory_code] = Time(
+            observations[observations["observatory_code"].isin([observatory_code])][
+                "mjd_utc"
+            ].unique(),
+            scale="utc",
+            format="mjd",
+        )
+
+    # Group the predicted ephemerides and observations by visit / exposure
+    observations_grouped = observations.groupby(by=["observatory_code", "mjd_utc"])
+    observations_visits = [
+        observations_grouped.get_group(g) for g in observations_grouped.groups
+    ]
+
+    # We pre-computed the ephemerides. Now we filter the ephemeris for only visits
+    # that have observations in the obs group passed to the worker
+
+    ephemeris_pre_grouped = ephemeris.groupby(by=["observatory_code", "mjd_utc"])
+    obs_group_keys = list(observations_grouped.groups.keys())
+    indices_to_drop = pd.Index([], dtype="int64")
+    for g_key in list(ephemeris_pre_grouped.groups.keys()):
+        if g_key not in obs_group_keys:
+            indices_to_drop = indices_to_drop.union(
+                ephemeris_pre_grouped.get_group(g_key).index
+            )
+
+    ephemeris_filtered = ephemeris.drop(indices_to_drop)
+
+    # Group the now-filtered ephemerides. There should only be visits for the observation set
+    ephemeris_grouped = ephemeris_filtered.groupby(by=["observatory_code", "mjd_utc"])
+    ephemeris_visits = [
+        ephemeris_grouped.get_group(g) for g in ephemeris_grouped.groups
+    ]
+
+    # Loop through each unique exposure and visit, find the nearest observations within
+    # tolerance (haversine metric)
+    distances = []
+    orbit_ids_associated = []
+    obs_ids_associated = []
+    mag_associated = []
+    mag_sigma_associated = []
+    obs_times_associated = []
+    coords_associated = []
+    coords_pred_associated = []
+    velocities_associated = []
+    coords_sigma_associated = []
+    frame_metadata_associated = []
+    tolerance_rad = np.radians(tolerance)
+    residuals = []
+    stats = []
+    for ephemeris_visit, observations_visit in zip(
+        ephemeris_visits, observations_visits
+    ):
+        assert len(ephemeris_visit["mjd_utc"].unique()) == 1
+        assert len(observations_visit["mjd_utc"].unique()) == 1
+        assert (
+            observations_visit["mjd_utc"].unique()[0]
+            == ephemeris_visit["mjd_utc"].unique()[0]
+        )
```

Comment:
This is where a major refactor takes place in the case of per-obs mjds. Not sure the BallTree implementation makes sense in the context where each obs has its own `coords_predicted_latlon`. Not much of a tree in that case.